Make table deletion a schema change so that truncation doesn't race with other people still using the table

andreimatei opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5907)

<!-- Reviewable:end -->

andreimatei committed:
sql: don't hand out leases for deleted tables

Fixes #5483
Fixes #5484

Also eliminates a race in leasing:
LeaseManager.RefreshLease() was reading tableState.active without
locking.

andreimatei committed:
sql: test that deleted table cannot be used. This failed before the previous fix

tschottdorf commented on sql/lease.go:
```diff
@@ -213,9 +213,15 @@ func (s LeaseStore) waitForOneVersion(tableID ID, retryOpts retry.Options) (Desc
-// called multiple times if retries occur: make sure it does not have side
-// effects.
+// Publish updates a table descriptor. It also maintains the invariant that
+// there's at most two versions of the descriptor out in the wild at any time
```
`there are`

tschottdorf commented on sql/lease.go:
```diff
@@ -213,9 +213,15 @@ func (s LeaseStore) waitForOneVersion(tableID ID, retryOpts retry.Options) (Desc
+// there's at most two versions of the descriptor out in the wild at any time
+// by first waiting for all nodes to be on the current (pre-update) version of
+// the table desc.
+// The update closure is called after the wait, and it's supposed to be the one
```
`after the wait, and provides the new version of the descriptor to be written`.

tschottdorf commented on sql/table.go:
```diff
@@ -43,6 +43,10 @@ func TestDisableTableLeases() func() {
 	}
 }
 
+func tableDoesntExistUErr(name string) *roachpb.Error {
```
`s/Doesnt/DoesNot/`?

tschottdorf commented on sql/table.go:
```diff
@@ -43,6 +43,10 @@ func TestDisableTableLeases() func() {
 }
 
+func tableDoesntExistUErr(name string) *roachpb.Error {
+	return roachpb.NewUErrorf("table %q does not exist", name)
```
can't use `fmt.Errorf` here and convert upstream (not necessary, just nice to push the `*Error` plague up as high as possible)

tschottdorf commented on sql/table.go:
```diff
@@ -199,23 +203,53 @@ func makeColumnDefDescs(d *parser.ColumnTableDef) (*ColumnDescriptor, *IndexDesc
+
+// getTableDescEx returns a table descriptor, or nil if the descriptor is not
+// found.
+func (p *planner) getTableDescEx(qname *parser.QualifiedName) (
```
what's the `Ex` suffix?

tschottdorf replied:
Why not mimick the interface of `p.getDescriptor`? Seems the right thing for the job.

tschottdorf commented on sql/table.go:
```diff
@@ -225,7 +259,9 @@ func getTableDescFromID(txn *client.Txn, id ID) (*TableDescriptor, *roachpb.Erro
 	if tableDesc == nil {
-		return nil, roachpb.NewErrorf("ID %d is not a table", id)
+		// TODO(andrei): We're theoretically hiding the error where desc exists, but
+		// is some other type of descriptor other than a TableDescriptor.
```
grammar.

vivekmenezes commented:
Review status: 0 of 31 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[sql/structured.proto, line 259 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEl94X96Y8F6Mzfjpzr:-KEl94X96Y8F6Mzfjpzs:-1247712585) ([raw file](https://github.com/cockroachdb/cockroach/blob/66f06cd0cef731839c93ec9a3c874bd151a51b7e/sql/structured.proto#L259)):*
I see that you have added the deleted bit in here as a mutation. Is there a reason why the deleted bit is not a part of the table descriptor itself? The reason I ask is because I do not like the descriptor is missing comment you have above. I'd much prefer deletion not be a mutation, because in my mind it's not a mutation to the descriptor. We just want to delete the descriptor.

Can you just delete all mutations and mark the table descriptor as deleted? That will disable all the mutation related work going on the table, and get it ready for deletion.

---

*[sql/parser/expr.go, line 536 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEl71BVvj9gE8qXkxHQ:-KEl71BVvj9gE8qXkxHR:1515906317) ([raw file](https://github.com/cockroachdb/cockroach/blob/66f06cd0cef731839c93ec9a3c874bd151a51b7e/sql/parser/expr.go#L536)):*
And if the qualified name is not a table name?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
sql: Make DROP TABLE be a schema change

On DROP TABLE, we now synchronously mark the table as DELETED and set
up_version. The schema changer will then publish this change. This
ensures (in conjunction with the next commit) that leases will not be
handed out any more.  Then, we asynchronously truncate and delete the
table, now that nobody's writing on it any more.

andreimatei committed:
sql: a more flexible mechanism for blocking schema changers in tests

Before this we have DisableSyncSchemaChanges, but that guy can only be
called before the server is created (otherwise it races with schema
changers). I need to be able to let schema changes run in the beginning
of the test, and block them later on. This will be used in the next
commit.

andreimatei committed:
sql: test that deleted tables cannot be leased

andreimatei commented:
@vivekmenezes I've removed the mutation. I've also added some other utility commits. PTAL when you can.

---

Review status: 0 of 35 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[sql/structured.proto, line 259 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEl94X96Y8F6Mzfjpzr:-KEquSfsBkhj2IUBarcN:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/66f06cd0cef731839c93ec9a3c874bd151a51b7e/sql/structured.proto#L259)):*
done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-:-KEquWeGO9d235f8M1q5:1624310144)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Thanks for making this change. I do like it but there is one more improvement worth making here or creating an issue over.

Once the table descriptor has been marked deleted and there are no more leases on the previous version, it would be nice to rename the table to _GC_<name> (or equivalent) so that going forward all schema change commands and other commands like creating a table with the same name behave nicely. If you have a very large table you need not have to wait until the table is completely GC-ed before you can reuse the name.

The above change will also make the thinking around everything else a lot easier. You don't have to worry about new schema change mutations being added for instance to the old descriptor.

---

Review status: 0 of 35 files reviewed at latest revision, 30 unresolved discussions, some commit checks failed.

---

*[roachpb/errors.go, line 556 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEr752xNFSo1ML8X_tG:-KEr752xNFSo1ML8X_tH:1172588067) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/roachpb/errors.go#L556)):*
return "descriptor deleted"

---

*[roachpb/errors.proto, line 193 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEr7axjQolHjzixp7mk:-KEr7axjQolHjzixp7ml:-1541105042) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/roachpb/errors.proto#L193)):*
is there value to having the id in DescriptorDeletedError as well?

---

*[sql/alter_table.go, line 182 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEr8Kw-tuE9a8MxXMiK:-KEr8Kw-tuE9a8MxXMiL:-1014583425) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/alter_table.go#L182)):*
See Ben's recent change. This is going to clash with that.

---

*[sql/descriptor.go, line 133 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEr9Y_NX70kBGw6q4Ed:-KEr9Y_NX70kBGw6q4Ee:39446883) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/descriptor.go#L133)):*
why named return values? If it's for the comment above just say return false, etc

---

*[sql/drop.go, line 172 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErBF8gPxejxgwlF2vy:-KErBF8gPxejxgwlF2vz:-529400255) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/drop.go#L172)):*
for _, name := range n.Names {

---

*[sql/drop.go, line 228 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErBrjwrt8iwXbgJ6jJ:-KErBrjwrt8iwXbgJ6jK:822322108) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/drop.go#L228)):*
You really don't need a new mutationID here. See what's done in Rename Index

---

*[sql/lease.go, line 493 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErCvGwYXOoPfTvEe8Z:-KErCvGwYXOoPfTvEe8_:997587217) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/lease.go#L493)):*
Why do we need this?

---

*[sql/lease.go, line 571 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErDiCpAmSMrR0eelis:-KErDiCpAmSMrR0eelit:103003808) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/lease.go#L571)):*
Not if it's a transaction that spans many requests.

---

*[sql/lease.go, line 667 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErEylT4a-1TNU0KpUA:-KErEylT4a-1TNU0KpUB:-1949881072) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/lease.go#L667)):*
Perhaps we should have the calls on one table descriptor cleanup the other tables or something like that.

---

*[sql/lease_test.go, line 348 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErGU6euhMNu1bl35DG:-KErGU6euhMNu1bl35DH:1649006699) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/lease_test.go#L348)):*
better to return an error

---

*[sql/lease_test.go, line 393 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErGGXpNEK1rKYtE6zd:-KErGGXpNEK1rKYtE6ze:1743013559) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/lease_test.go#L393)):*
Exec()

---

*[sql/lease_test.go, line 405 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErGi3W6mbV-2bFH1TG:-KErGi3W6mbV-2bFH1TH:1043274689) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/lease_test.go#L405)):*
Let's acquire a lease on the table descriptor here before we drop the table and release it so we have the cached version.

---

*[sql/lease_test.go, line 407 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErHCQNO1h91tNoQubw:-KErHCQNO1h91tNoQubx:1743013559) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/lease_test.go#L407)):*
Exec()

---

*[sql/lease_test.go, line 418 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErFiofy0Cc0xXsi5pp:-KErFiofy0Cc0xXsi5pq:577013972) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/lease_test.go#L418)):*
place on line above

---

*[sql/plan.go, line 389 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErIg18qBgTBq3Opfmx:-KErIg18qBgTBq3Opfmy:2117603263) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/plan.go#L389)):*
No good reason for this tmp

---

*[sql/schema_changer.go, line 65 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErJ5NMZ6qiVyUkJxAn:-KErJ5NMZ6qiVyUkJxAo:-1734305303) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/schema_changer.go#L65)):*
It will be nice to not use a planner since we already dropped it from  backfill

---

*[sql/schema_changer.go, line 90 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErJIuihZnxjj47Tmv5:-KErJIuihZnxjj47Tmv6:-1674607694) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/schema_changer.go#L90)):*
please add this line to the getTableDescriptorAtVersion() function after rebasing

---

*[sql/schema_changer.go, line 142 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErJl8ElhD_2kWZsd9x:-KErJl8ElhD_2kWZsd9y:-1954154706) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/schema_changer.go#L142)):*
I think it's best to not add the error all over the place. Look at my comment below.

---

*[sql/schema_changer.go, line 246 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErKfufGSVDrYJpFYqj:-KErKfufGSVDrYJpFYqk:-2096238761) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/schema_changer.go#L246)):*
Wanna add a comment like:

From this point forward it is guaranteed that no node will change the underlying data on the table with the exception of the backfill.  It is important to grab a schema change lease on the table now to prevent some backfill operation from adding more data. Schema change commands can continue executing on the table descriptor. That's fine, but we don't want any backfill being triggered. I suppose that's protected by the fact that the backfill checks whether the descriptor has been deleted.

---

*[sql/schema_changer.go, line 652 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErML3pM_hh12kowiKC:-KErML3pM_hh12kowiKD:1098102854) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/schema_changer.go#L652)):*
No. The schema change will not be present on the table descriptor in the kv store.

---

*[sql/structured.proto, line 257 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEr58Ja1cXpNCTfDlOC:-KEr58Jb2Sjpk7ZCfPVc:1015197542) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/structured.proto#L257)):*
It is illegal to transition deleted = true to deleted = false

---

*[sql/parser/expr.go, line 540 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEr5qT5XFAikeorqoS_:-KEr5qT5XFAikeorqoSa:160365993) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/parser/expr.go#L540)):*
> 0

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-:-KErPWDDNS24CTPGi239:-1905995702)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
The more i think about this, lets make drop table super fast and decouple
the data deletion from it by gcing the table data in the async schema
changer.

I'm happy to chat about this before you proceed

On Fri, Apr 8, 2016, 1:29 PM Andrei Matei <notifications@github.com> wrote:

> @vivekmenezes <https://github.com/vivekmenezes> I've removed the
> mutation. I've also added some other utility commits. PTAL when you can.
> ------------------------------
>
> Review status: 0 of 35 files reviewed at latest revision, 8 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *sql/structured.proto, line 259 [r5]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEl94X96Y8F6Mzfjpzr:-KEquSfsBkhj2IUBarcN:677588487>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/66f06cd0cef731839c93ec9a3c874bd151a51b7e/sql/structured.proto#L259>):*
> done
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-:-KEquWeGO9d235f8M1q5:1624310144>*
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/5907#issuecomment-207526851>
>

petermattis commented:
We even have an issue filed to optimize DROP/TRUNCATE:
https://github.com/cockroachdb/cockroach/issues/2003.

On Fri, Apr 8, 2016 at 4:01 PM, vivekmenezes <notifications@github.com>
wrote:

> The more i think about this, lets make drop table super fast and decouple
> the data deletion from it by gcing the table data in the async schema
> changer.
>
> I'm happy to chat about this before you proceed
>
> On Fri, Apr 8, 2016, 1:29 PM Andrei Matei <notifications@github.com>
> wrote:
>
> > @vivekmenezes <https://github.com/vivekmenezes> I've removed the
> > mutation. I've also added some other utility commits. PTAL when you can.
> > ------------------------------
> >
> > Review status: 0 of 35 files reviewed at latest revision, 8 unresolved
> > discussions, some commit checks failed.
> > ------------------------------
> >
> > *sql/structured.proto, line 259 [r5]
> > <
> https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEl94X96Y8F6Mzfjpzr:-KEquSfsBkhj2IUBarcN:677588487
> >
> > (raw file
> > <
> https://github.com/cockroachdb/cockroach/blob/66f06cd0cef731839c93ec9a3c874bd151a51b7e/sql/structured.proto#L259
> >):*
> > done
> > ------------------------------
> >
> > *Comments from Reviewable
> > <
> https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-:-KEquWeGO9d235f8M1q5:1624310144
> >*
> >
> > —
> > You are receiving this because you were mentioned.
> >
> >
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/cockroachdb/cockroach/pull/5907#issuecomment-207526851>
>
> >
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/5907#issuecomment-207579770>
>

andreimatei commented:
Review status: 0 of 35 files reviewed at latest revision, 30 unresolved discussions, some commit checks failed.

---

*[roachpb/errors.go, line 556 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEr752xNFSo1ML8X_tG:-KFR84VRfOTNPa4mY7ZD:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/roachpb/errors.go#L556)):*
Done.

---

*[roachpb/errors.proto, line 193 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEr7axjQolHjzixp7mk:-KFR8EGGvCgUoAhm9sYw:-2026230130) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/roachpb/errors.proto#L193)):*
i don't think so. That's an internal error, we never print it or return it to the user.

---

*[sql/descriptor.go, line 133 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEr9Y_NX70kBGw6q4Ed:-KFR2FIfLPwHDyooPfwH:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/descriptor.go#L133)):*
Done.

---

*[sql/drop.go, line 172 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErBF8gPxejxgwlF2vy:-KFR1AG-Dt3ifLluB1tv:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/drop.go#L172)):*
Done.

---

*[sql/drop.go, line 228 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErBrjwrt8iwXbgJ6jJ:-KFQKteNvy7z34llt9P8:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/drop.go#L228)):*
Done.

---

*[sql/lease.go, line 217 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEkKKL1DxmML3fmvK16-r2-217:-KFR1URFfjX_TG5N2HaN:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/679070fce939fffd9da36ab5a7b2d66854944cd4/sql/lease.go#L217)):*
Done.

---

*[sql/lease.go, line 220 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEkKKL1DxmML3fmvK16-r2-220:-KFR1_uj00lTvKWmv3eF:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/679070fce939fffd9da36ab5a7b2d66854944cd4/sql/lease.go#L220)):*
Done.

---

*[sql/lease.go, line 493 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErCvGwYXOoPfTvEe8Z:-KFR4rUzB48EqmsmzmTI:-583328562) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/lease.go#L493)):*
well it's needed because that's this `acquire` is how leases are refreshed. If we just return here, the `releaseNonLatest` below doesn't run and so we never release leases for deleted tables.

---

*[sql/lease_test.go, line 348 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErGU6euhMNu1bl35DG:-KFR5qsSHlBVw4m0v8kT:-600357) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/lease_test.go#L348)):*
meh, it's a test

---

*[sql/lease_test.go, line 393 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErGGXpNEK1rKYtE6zd:-KFR5xleqkNQlgYZkg5_:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/lease_test.go#L393)):*
Done.

---

*[sql/lease_test.go, line 405 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErGi3W6mbV-2bFH1TG:-KFR6ZbE-A8pztQeq5lX:-1617690340) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/lease_test.go#L405)):*
That's not so easy, since there's no releasing happening in this test (I blocked the schema changers from running above because otherwise I can't reliably perform an acquire while the descriptor is in the deleted state, as opposed to actually deleted). This is also the reason why we do acquire with a bogus node id below.

---

*[sql/lease_test.go, line 407 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErHCQNO1h91tNoQubw:-KFR63_jb6hCWS4VGrq5:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/lease_test.go#L407)):*
Done.

---

*[sql/lease_test.go, line 418 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErFiofy0Cc0xXsi5pp:-KFR6AacmOkqAEDA29Ac:-78967346) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/lease_test.go#L418)):*
doesn't fit

---

*[sql/plan.go, line 389 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErIg18qBgTBq3Opfmx:-KFR1uty5hZOn9LdREul:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/plan.go#L389)):*
Done.

---

*[sql/schema_changer.go, line 65 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErJ5NMZ6qiVyUkJxAn:-KFQwXbpIjilq3dGLILy:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/schema_changer.go#L65)):*
Done.

---

*[sql/schema_changer.go, line 90 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErJIuihZnxjj47Tmv5:-KFQl_PWoWe7s8IoXCh-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/schema_changer.go#L90)):*
Done.

---

*[sql/schema_changer.go, line 142 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErJl8ElhD_2kWZsd9x:-KFQzn2mw-wZLkDkf9UO:-789222242) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/schema_changer.go#L142)):*
sorry, which comment is that?
This error is useful here for informing the async schema changers runner that it can drop the schema changer for this deletion. Otherwise, nothing would remove it from the list of schema changers to run.

---

*[sql/schema_changer.go, line 246 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErKfufGSVDrYJpFYqj:-KFQzShQhNWt5CleLbb7:-1333595122) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/schema_changer.go#L246)):*
expanded the comment above a bit.
I don't think it's worth complicating the comment further, talking about backfill. Indeed, backfill doesn't run if the table is found to have `deleted` set.

---

*[sql/schema_changer.go, line 652 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErML3pM_hh12kowiKC:-KFQ7LmvIjQFeSiHYE9T:160401764) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/schema_changer.go#L652)):*
ack

---

*[sql/structured.proto, line 257 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEr58Ja1cXpNCTfDlOC:-KFR02oNOzGZM2tRk5Ht:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/structured.proto#L257)):*
Done.

---

*[sql/table.go, line 46 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEkKKL1DxmML3fmvK19-r2-46:-KFQ0K7IU7Yb5y51HH-k:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/679070fce939fffd9da36ab5a7b2d66854944cd4/sql/table.go#L46)):*
Done.

---

*[sql/table.go, line 47 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEkKKL1DxmML3fmvK19-r2-47:-KFQ0KZHfl2WUqfzNUOv:-1201783071) ([raw file](https://github.com/cockroachdb/cockroach/blob/679070fce939fffd9da36ab5a7b2d66854944cd4/sql/table.go#L47)):*
meh. This function is only used in a handful of places that want perr

---

*[sql/table.go, line 217 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEkKKL1DxmML3fmvK19-r2-217:-KFPpZD6c8Sjwj3PRNuo:-289024313) ([raw file](https://github.com/cockroachdb/cockroach/blob/679070fce939fffd9da36ab5a7b2d66854944cd4/sql/table.go#L217)):*
"Ex" is extended :). Win32 API convention, I think.

Why do you like `p.getDescriptor()` (i.e. having the caller pass in a proto to be populated versus returning a (pointer to) proto)? To save the proto allocation? In fact, I think we should change that interface.

---

*[sql/table.go, line 263 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEkKKL1DxmML3fmvK19-r2-263:-KFPqbXSLdUxa_Vekr9k:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/679070fce939fffd9da36ab5a7b2d66854944cd4/sql/table.go#L263)):*
Done.

---

*[sql/parser/expr.go, line 536 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEl71BVvj9gE8qXkxHQ:-KFPrXLI03Fs9hZgA6v-:296107657) ([raw file](https://github.com/cockroachdb/cockroach/blob/66f06cd0cef731839c93ec9a3c874bd151a51b7e/sql/parser/expr.go#L536)):*
if it's not a table name, you get something bogus. Garbage in, garbage out. Added a comment.
Or are you suggesting putting it some checks like in `NormalizeTableName()`? I think that'd be overkill.

---

*[sql/parser/expr.go, line 540 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KEr5qT5XFAikeorqoS_:-KFPsMtH0tdMM3INGkg6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/parser/expr.go#L540)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
@vivekmenezes as we discussed, let's do async truncate in a future PR

---

Review status: 0 of 35 files reviewed at latest revision, 30 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-:-KFRA7ghS3Nu6x4-ic4C:869005657)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 35 files reviewed at latest revision, 37 unresolved discussions, some commit checks failed.

---

*[sql/backfill.go, line 104 \[r22\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFeHwbsUjnBY7yL4D94:-KFeHwbtjmX0hpQj_foZ:1202062178) ([raw file](https://github.com/cockroachdb/cockroach/blob/2387ddd5916a62c307eae82ed54d9fb2fcf0c90d/sql/backfill.go#L104)):*
How can this happen. Shouldn't getTableDescFromID() return an error?

---

*[sql/config.go, line 67 \[r22\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFeIFv1Bp2ZdsPCmR2u:-KFeIFv1Bp2ZdsPCmR2v:-1595431322) ([raw file](https://github.com/cockroachdb/cockroach/blob/2387ddd5916a62c307eae82ed54d9fb2fcf0c90d/sql/config.go#L67)):*
probably makes sense for it to return an error?

---

*[sql/lease.go, line 493 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErCvGwYXOoPfTvEe8Z:-KFeN90BJEwOOye-sg6S:803644733) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/lease.go#L493)):*
Instead of doing this here, can we capture all this special logic in refreshMaybe (refreshLeases) so that the code path called from the async path is the only one releasing the old leases on the table.

---

*[sql/lease.go, line 130 \[r22\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFeKbU4ojzCZKY5Bguk:-KFeKbU4ojzCZKY5Bgul:-254518893) ([raw file](https://github.com/cockroachdb/cockroach/blob/2387ddd5916a62c307eae82ed54d9fb2fcf0c90d/sql/lease.go#L130)):*
move this above the version check above

---

*[sql/lease_test.go, line 411 \[r22\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFeNr5QoSbmi8QE4oqW:-KFeNr5QoSbmi8QE4oqX:2013915209) ([raw file](https://github.com/cockroachdb/cockroach/blob/2387ddd5916a62c307eae82ed54d9fb2fcf0c90d/sql/lease_test.go#L411)):*
Can you use the same logic used in TestRaceWithBackfill which runs the schema change in a goroutine and waits for a particular version after which all leases will be denied. I'd also add some rows into the table so that the truncation takes some time.

---

*[sql/schema_changer.go, line 142 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErJl8ElhD_2kWZsd9x:-KFePYG5HubNQQOSo8Wl:-1360686922) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/schema_changer.go#L142)):*
Actually leave this out and create a new issue for me. I'd rather you merge this in in a state where the truncation logic starts running after all the mutations have been processed. So the deletion on its own for now will not delete the outstanding mutations. In a separate change I'm making a number of changes and I can see how I can add in the mutation deletion logic better without requiring these changes. Create an issue for me "speed up table drop by disabling schema changers before truncating table"

---

*[sql/schema_changer.go, line 191 \[r22\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFeRelXsCMewNWS-sAJ:-KFeRelXsCMewNWS-sAK:-502124565) ([raw file](https://github.com/cockroachdb/cockroach/blob/2387ddd5916a62c307eae82ed54d9fb2fcf0c90d/sql/schema_changer.go#L191)):*
For now truncate and drop the table if there are no outstanding mutations on the table.

---

*[sql/schema_changer.go, line 549 \[r22\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFeSVABB_0uVkmWmT1F:-KFeSVABB_0uVkmWmT1G:258287067) ([raw file](https://github.com/cockroachdb/cockroach/blob/2387ddd5916a62c307eae82ed54d9fb2fcf0c90d/sql/schema_changer.go#L549)):*
you also need to enqueue for table deletion. Remember Version++ and table deletion happen in two separate transactions, a schema changer can run version++ and then fail resulting in the table never getting deleted. Add a comment here to that effect.

---

*[sql/structured.go, line 642 \[r22\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFeU45dXPXSEMY1a-WF:-KFeU45dXPXSEMY1a-WG:188580534) ([raw file](https://github.com/cockroachdb/cockroach/blob/2387ddd5916a62c307eae82ed54d9fb2fcf0c90d/sql/structured.go#L642)):*
I actually do not like the use a boolean here. I think there is value to having two separate methods

setUpVersion() // only sets upversion

newMutationID()  // gets new mutation id and sets upversion.

Add a TODO to eliminate newMutationID()  in favor of appendMutation() or something more sensible.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 35 files reviewed at latest revision, 37 unresolved discussions, some commit checks failed.

---

*[sql/backfill.go, line 104 \[r22\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFeHwbsUjnBY7yL4D94:-KFf7G1JqChjtN4N2gzu:1291674890) ([raw file](https://github.com/cockroachdb/cockroach/blob/2387ddd5916a62c307eae82ed54d9fb2fcf0c90d/sql/backfill.go#L104)):*
let's leave it retuning nil. It's convenient for `SchemaChangeManager.IsDone`

---

*[sql/config.go, line 67 \[r22\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFeIFv1Bp2ZdsPCmR2u:-KFf80PC0DSzeVk5T_SG:102821695) ([raw file](https://github.com/cockroachdb/cockroach/blob/2387ddd5916a62c307eae82ed54d9fb2fcf0c90d/sql/config.go#L67)):*
`GetZoneConfig` above also returns nil. I think that's better than an error (easier to test for).

---

*[sql/lease.go, line 493 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErCvGwYXOoPfTvEe8Z:-KFfEGBOaXUZ12_e_6Af:1540687024) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/lease.go#L493)):*
OK, moved to refreshMaybe

---

*[sql/lease.go, line 130 \[r22\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFeKbU4ojzCZKY5Bguk:-KFf8VRZGPrCy6mKO0JU:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2387ddd5916a62c307eae82ed54d9fb2fcf0c90d/sql/lease.go#L130)):*
Done.

---

*[sql/lease_test.go, line 411 \[r22\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFeNr5QoSbmi8QE4oqW:-KFf9HAXHVe-R5PHcpK0:631488928) ([raw file](https://github.com/cockroachdb/cockroach/blob/2387ddd5916a62c307eae82ed54d9fb2fcf0c90d/sql/lease_test.go#L411)):*
but in this test we don't want to run any schema changes. We just mark the table for deletion and verify that random nodes can't get a lease on it (so there's no `Publish` call).

---

*[sql/schema_changer.go, line 142 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KErJl8ElhD_2kWZsd9x:-KFelgBXtxsBQZHI6mJ2:2096837899) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa1afca8ae701afd24ba0b235da485cfbb9a923/sql/schema_changer.go#L142)):*
as we discussed, this is still needed

---

*[sql/schema_changer.go, line 191 \[r22\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFeRelXsCMewNWS-sAJ:-KFen38YKNtD5LdLvRF8:-226086928) ([raw file](https://github.com/cockroachdb/cockroach/blob/2387ddd5916a62c307eae82ed54d9fb2fcf0c90d/sql/schema_changer.go#L191)):*
I tried to do it but it leads to pretty ugly code, duplicating some call to `sc.truncateAndDropTable()` both in the short-circuit below for `invalidMutationID` and at the end of the method.
Besides, we can in fact run the deletion even if there are queued mutations, so I'd leave it the way it is for now.

---

*[sql/schema_changer.go, line 549 \[r22\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFeSVABB_0uVkmWmT1F:-KFejNllw02sESY6bgCS:801661331) ([raw file](https://github.com/cockroachdb/cockroach/blob/2387ddd5916a62c307eae82ed54d9fb2fcf0c90d/sql/schema_changer.go#L549)):*
done, good catch

---

*[sql/structured.go, line 642 \[r22\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFeU45dXPXSEMY1a-WF:-KFf6Nu71wQvR7XdC6OG:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2387ddd5916a62c307eae82ed54d9fb2fcf0c90d/sql/structured.go#L642)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
I've taken another pass at this. Most of the comments are in the lease and schema change area, and very little is in the rest of the refactoring. It would make me very happy if you can take lease and schema changer out of this change and do the basic refactoring stuff in a separate PR, or consolidate this entire PR into three commits
1. refactoring, 2. leases, 3. schema change changes.

Thanks!

---

Review status: 0 of 35 files reviewed at latest revision, 49 unresolved discussions, some commit checks failed.

---

*[sql/drop_test.go, line 284 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjGSRzHMk7QXNhTxmT:-KFjGSS-gpGMpFpfk0jL:-1760146717) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/drop_test.go#L284)):*
Can you add some more data into the table and then make this test similar to TestRaceWithBackfill where DROP TABLE is called in a goroutine and the below SELECT is called immediately after the version has incremented. Thanks

---

*[sql/lease.go, line 492 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjLmSxrM5bJIC_9VNu:-KFjLmSxrM5bJIC_9VNv:-1460838858) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/lease.go#L492)):*
As I said earlier I don't believe you need to purge the old leases here. Just do them in purgeOldLeases called from the async path.

---

*[sql/lease.go, line 496 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjID57NNikItQi4mnw:-KFjID58ePyZ_ipRsdHl:-212323963) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/lease.go#L496)):*
Why did you remove the call to releaseNonlatest() below? I suppose you want to depend on the async path. I think it's okay to take it out.

---

*[sql/lease.go, line 522 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjMimrhmo6sCnIxraZ:-KFjMimrhmo6sCnIxra_:-213755093) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/lease.go#L522)):*
Add a comment that this method depends on there being at least one active lease

---

*[sql/lease.go, line 590 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjLAB6TUOXn3WV8Fu0:-KFjLAB6TUOXn3WV8Fu1:209043265) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/lease.go#L590)):*
I think we should rename this method to purgeOldLeases()

---

*[sql/lease.go, line 598 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjLXz1ACjHouu3PBsf:-KFjLXz1ACjHouu3PBsg:611739249) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/lease.go#L598)):*
No strong reason to define this variable

---

*[sql/lease.go, line 613 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjL2NAGqpAI6SEonkY:-KFjL2NAGqpAI6SEonkZ:-1346741013) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/lease.go#L613)):*
I'm very confused by this comment. Why can't you use Acquire?

---

*[sql/lease.go, line 626 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjKUrNE8N1ign8SE21:-KFjKUrNE8N1ign8SE22:-695646423) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/lease.go#L626)):*
I'd swallow the DescriptorDeletedError here

---

*[sql/lease.go, line 636 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjKl2AvoxbpJGkTB49:-KFjKl2AvoxbpJGkTB4A:2005737700) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/lease.go#L636)):*
I don't think we need this extra call to release

---

*[sql/schema_changer.go, line 171 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjOZLxOHn-RTILtCLA:-KFjOZLxOHn-RTILtCLB:1466142002) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/schema_changer.go#L171)):*
Do you really need this needRelease stuff. Can ReleaseLease return an error instead.

Perhaps what you want is to not run the truncate and delete code through exec() and instead call its own method deleteTable(). See purgeMutations().

---

*[sql/schema_changer.go, line 178 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjPekaPrupfKsVKnat:-KFjPekaPrupfKsVKnau:-1629931978) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/schema_changer.go#L178)):*
I see that you changed the signature of Publish() to return the table descriptor. Is there a strong reason to do that?

---

*[sql/schema_changer.go, line 598 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjQ4FLivQSZuVYNQ9K:-KFjQ4FLivQSZuVYNQ9L:-1735614572) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/schema_changer.go#L598)):*
Interesting. Are you sure about that? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-:-KFjQs_d-AlJ0vR2pvuL:912902237)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
@vivekmenezes the other refactoring PR has gone in. Mind taking another look?
You can now even "combine commits for review" and it's not so bad :). Thanks!

---

Review status: 0 of 26 files reviewed at latest revision, 49 unresolved discussions, some commit checks pending.

---

*[sql/drop_test.go, line 284 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjGSRzHMk7QXNhTxmT:-KFk60QhE1bMA4TofgWm:646185238) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/drop_test.go#L284)):*
but what's the point?  Isn't it a good thing that we test the sync schema changer? This is the way we want the workflow to generally work - user DROPs and then she immediately fails to use the table, in the same session.

---

*[sql/lease.go, line 492 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjLmSxrM5bJIC_9VNu:-KFkNgVKO4C8QxOcXLNH:1216187601) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/lease.go#L492)):*
OK you're right, lease releasing can be done exclusively in the gossip handler

---

*[sql/lease.go, line 496 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjID57NNikItQi4mnw:-KFkNhv02j5REygD58t5:-1860931759) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/lease.go#L496)):*
right, now that async path is the only one doing releases

---

*[sql/lease.go, line 522 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjMimrhmo6sCnIxraZ:-KFkDOuhTBx-7FfDh3MG:22985368) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/lease.go#L522)):*
does it depend on that?

---

*[sql/lease.go, line 590 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjLAB6TUOXn3WV8Fu0:-KFkORewr624w99bggts:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/lease.go#L590)):*
Done.

---

*[sql/lease.go, line 598 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjLXz1ACjHouu3PBsf:-KFkOdv1M8eMAVbMhEN1:1743906955) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/lease.go#L598)):*
isn't the inline function bad enough? You want to declare it in an `if`?

---

*[sql/lease.go, line 613 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjL2NAGqpAI6SEonkY:-KFk_qKwkrdiCIQl1t8Q:-1811522747) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/lease.go#L613)):*
comment gone

---

*[sql/lease.go, line 626 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjKUrNE8N1ign8SE21:-KFk_sFuIVVJbF8CszPp:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/lease.go#L626)):*
Done.

---

*[sql/lease.go, line 636 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjKl2AvoxbpJGkTB49:-KFka0RyE_MPUAa7x8M3:1789708697) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/lease.go#L636)):*
I don't think it's "extra", I think it's the only call that releases the lease taken on line 613 above

---

*[sql/schema_changer.go, line 171 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjOZLxOHn-RTILtCLA:-KFkaVvo24hJYCIUaJek:1230811207) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/schema_changer.go#L171)):*
it's needed because we want to short-circuit `sc.ReleaseLease()` below. Even if deletion was its own function (and in fact it pretty much is all contained in `truncateAndDropTable`), we'd still need to short-circuit that call. Unless we call `truncateAndDropTable` before the call to `sc.AcquireLease` above, but then we'd have to copy over the `MaybeIncrementVersion` call too...

---

*[sql/schema_changer.go, line 178 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjPekaPrupfKsVKnat:-KFkbMf1B7adPMcm2EiA:547242872) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/schema_changer.go#L178)):*
discussed offline - seems like a natural API

---

*[sql/schema_changer.go, line 598 \[r30\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-KFjQ4FLivQSZuVYNQ9K:-KFkbXEPAQBOFjxAadiP:-1748428865) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e1487a00f0f6f2a34e6f60a2f8df47f9bb6c37a/sql/schema_changer.go#L598)):*
well, I guess we could trigger a gossip update by calling `txn.SetSystemConfigTrigger` in the txn that deletes the table descriptor, but that seems fragile and async and complex to reason about. What I'm doing here seems to me like a clean solution for cleaning up these schema changers.

<later>
In the meantime, I was forced to set `systemConfigTrigger` on that txn (because of a check we have in some tests that every KV transaction that touches system keys has that flag set), but I would still keep this code. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-:-KFqSOd3Wms_Wfi60Kr7:-1629470337)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
I'd like you to split this into four PRs:

"sql: add delete bit to table descriptor; schema change commands error on seeing the delete bit"

and another PR called

"sql: added TestingSchemaChangerCollection for  ..."

and another

"sql: ensure that a lease is not taken on a deleted table"

and 

"sql: delete table safely after all leases have expired"

I know this is more work but 1. you'll see these PRs getting checked in, and 2. I wont be sweating each time I have to review this.

Thanks

---

Review status: 0 of 26 files reviewed at latest revision, 45 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5907#-:-KFt8H0W7hEqucGQgWCQ:1824832411)*
<!-- Sent from Reviewable.io -->
